src/burn/engine/bundlepackageengine.cpp

```diff
@@ -92,45 +209,95 @@ extern "C" void BundlePackageEnginePackageUninitialize(
-    HKEY hkRoot = pPackage->fPerMachine ? HKEY_LOCAL_MACHINE : HKEY_CURRENT_USER;
-    REG_KEY_BITNESS bitness = pPackage->Bundle.fWin64 ? REG_KEY_64BIT : REG_KEY_32BIT;
-
-    // TODO: detect all related bundles, so that the Obsolete state can be detected.
-    hr = RegOpenEx(hkRoot, pPackage->Bundle.sczRegistrationKey, KEY_QUERY_VALUE, bitness, &hkRegistration);
-    if (SUCCEEDED(hr))
+    BUNDLE_QUERY_CONTEXT queryContext = { };
+
+    queryContext.pPackage = pPackage;
+    queryContext.pUserExperience = pUserExperience;
+
+    hr = BundleQueryRelatedBundles(
+        BUNDLE_INSTALL_CONTEXT_MACHINE,
+        const_cast<LPCWSTR*>(pPackage->Bundle.rgsczDetectCodes),
+        pPackage->Bundle.cDetectCodes,
+        const_cast<LPCWSTR*>(pPackage->Bundle.rgsczUpgradeCodes),
+        pPackage->Bundle.cUpgradeCodes,
+        const_cast<LPCWSTR*>(pPackage->Bundle.rgsczAddonCodes),
+        pPackage->Bundle.cAddonCodes,
+        const_cast<LPCWSTR*>(pPackage->Bundle.rgsczPatchCodes),
+        pPackage->Bundle.cPatchCodes,
+        QueryRelatedBundlesCallback,
+        &queryContext);
+    ExitOnFailure(hr, "Failed to query per-machine related bundle packages.");
+
+    hr = BundleQueryRelatedBundles(
+        BUNDLE_INSTALL_CONTEXT_USER,
+        const_cast<LPCWSTR*>(pPackage->Bundle.rgsczDetectCodes),
+        pPackage->Bundle.cDetectCodes,
+        const_cast<LPCWSTR*>(pPackage->Bundle.rgsczUpgradeCodes),
+        pPackage->Bundle.cUpgradeCodes,
+        const_cast<LPCWSTR*>(pPackage->Bundle.rgsczAddonCodes),
+        pPackage->Bundle.cAddonCodes,
+        const_cast<LPCWSTR*>(pPackage->Bundle.rgsczPatchCodes),
+        pPackage->Bundle.cPatchCodes,
+        QueryRelatedBundlesCallback,
+        &queryContext);
+    ExitOnFailure(hr, "Failed to query per-user related bundle packages.");
+
+    if (queryContext.fNewerFound)
     {
-        hr = RegReadNumber(hkRegistration, REGISTRY_BUNDLE_INSTALLED, &dwInstalled);
+        pPackage->currentState = queryContext.fSelfFound ? BOOTSTRAPPER_PACKAGE_STATE_SUPERSEDED : BOOTSTRAPPER_PACKAGE_STATE_OBSOLETE;
     }
-
-    // Not finding the key or value is okay.
-    if (E_FILENOTFOUND == hr || E_PATHNOTFOUND == hr)
+    else
     {
-        hr = S_OK;
+        pPackage->currentState = queryContext.fSelfFound ? BOOTSTRAPPER_PACKAGE_STATE_PRESENT : BOOTSTRAPPER_PACKAGE_STATE_ABSENT;
     }
 
-    fDetected = (1 == dwInstalled);
-
-    // update detect state
-    pPackage->currentState = fDetected ? BOOTSTRAPPER_PACKAGE_STATE_PRESENT : BOOTSTRAPPER_PACKAGE_STATE_ABSENT;
-
     if (pPackage->fCanAffectRegistration)
     {
         pPackage->installRegistrationState = BOOTSTRAPPER_PACKAGE_STATE_ABSENT < pPackage->currentState ? BURN_PACKAGE_REGISTRATION_STATE_PRESENT : BURN_PACKAGE_REGISTRATION_STATE_ABSENT;
     }
 
-    ReleaseRegKey(hkRegistration);
+    // TODO: The bundle is registering itself as a dependent when installed as a chain package, which prevents us from uninstalling it.
```

Comment:
I believe that the `-parent` switch is what's supposed to prevent the bundle from registering itself as a dependent. Can someone confirm that this is the correct usage of that switch? Or is this going to need to be a new feature because that switch is for something else?